core/src/main/java/edu/ucr/cs/riple/core/metadata/submodules/DownStreamDependencyAnalyzer.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2022 Nima Karimipour
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package edu.ucr.cs.riple.core.metadata.submodules;
+
+import com.google.common.collect.ImmutableSet;
+import edu.ucr.cs.riple.core.Config;
+import edu.ucr.cs.riple.core.Report;
+import edu.ucr.cs.riple.core.explorers.Explorer;
+import edu.ucr.cs.riple.core.explorers.OptimizedExplorer;
+import edu.ucr.cs.riple.core.injectors.VirtualInjector;
+import edu.ucr.cs.riple.core.metadata.field.FieldDeclarationAnalysis;
+import edu.ucr.cs.riple.core.metadata.index.Bank;
+import edu.ucr.cs.riple.core.metadata.index.Error;
+import edu.ucr.cs.riple.core.metadata.index.Fix;
+import edu.ucr.cs.riple.core.metadata.method.MethodInheritanceTree;
+import edu.ucr.cs.riple.core.metadata.method.MethodNode;
+import edu.ucr.cs.riple.core.metadata.trackers.MethodRegionTracker;
+import edu.ucr.cs.riple.core.util.Utility;
+import edu.ucr.cs.riple.injector.changes.AddAnnotation;
+import edu.ucr.cs.riple.injector.location.OnMethod;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * Analyzer for downstream dependencies. It collects public APIs used in the dependencies and
+ * collects a summarization of an overall effect across all downstream dependencies. These data can
```

Comment:
What do we mean here by "collects a summarization of an overall effect"? I thought effect in NullAwayInfer parlance was the change in the number of reported NullAway errors so... "collects the effect (number of additional errors) of the change by summing across all downstream dependencies", maybe? Is that what you mean?

Also, this doesn't collect the public APIs, does it? That's part of the Scanner. What this does is switch the nullability of those APIs to compute the effect on downstream dependencies of adding annotations to those APIs.

And, data is uncountable: "This data"...

Maybe, putting that all together:

> Analyzer for downstream dependencies.
> 
> This class analyzes switching the nullability of public APIs of one compilation target, in order to compute the effect on said target's downstream dependencies of adding annotations to those APIs. It does so by building said downstream dependencies. It collects the effect (number of additional errors) of each such change, by summing across all downstream dependencies. This data then be fed to the Annotator main process in the decision process.

Does that sound about right for what this class does?

Reply:
`Scanner` checker finds the list of all public methods in target module. This class runs the `Scanner` checker on dependencies and checks what subset of public methods are used in the dependency via `Trackers` and then does all the rest you mentioned in the suggested comment. I think I am just going to use what you mentioned here, since the part where unused methods are filtered is already described in the code itself.

7717c14628afbb0d49360dc793d5fcb3b18350ee